### PR TITLE
Adjust catch scoring to match stable score V2

### PR DESCRIPTION
**Pull request by bdach**
- Addresses https://github.com/ppy/osu/discussions/26040

This pull request changes the catch scoring algorithm to - as far as I can tell - match stable score V2 exactly.

### All non-bonus objects now are equally weighted for accuracy

[Compare stable.](https://github.com/peppy/osu-stable-reference/blob/master/osu!/GameplayElements/Scoring/ScoreFruits.cs#L34-L42)

This is achieved by assigning 300 base value to `Great`, `LargeTickHit`, and `SmallTickHit` in `GetBaseScoreForResult()`.

Notably, there are a [couple](https://github.com/ppy/osu/blob/fbc40ffc65930f1e36bb048c73623dbbd0e23062/osu.Game/Database/StandardisedScoreMigrationTools.cs#L56-L69) of [usages](https://github.com/ppy/osu-queue-score-statistics/blob/14e96bd47785823c9cdc50f6d5667db2e1f26db7/osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/BatchInserter.cs#L300) that sort by `GetBaseScoreForResult()`. The first I honestly can't really bring myself to care about too much (responsible for converting ancient lazer scores to new algorithm, which I find disposable and irrelevant at this time, but should be trivial to fix if we do care), and the second filters by `IsBasic()`, so is not broken by this in any capacity.

### Combo / "accuracy" apportioning is now beatmap-dependent

[Compare stable.](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreProcessorFruits.cs#L55-L58)

The "accuracy" proportion of score is given by

$$ 400\,000 \cdot \frac{\texttt{numSmallDroplets}}{\texttt{numSmallDroplets} + \texttt{numFruits}} $$

And yes, it is intentional that the denominator does not include large droplets. [Compare stable.](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreProcessor.cs#L34)

If the denominator of the expression above is 0, the accuracy portion is considered non-existent, which also matches stable.

The combo portion is 1 million minus the above.

### Only fruits and large droplets contribute to combo portion

[Compare stable.](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreProcessorFruits.cs#L66-L67)

Notably, for combo portion, normal fruits are weighted 300 and large droplets are weighted 100. [Compare stable.](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreChangeFruits.cs#L19-L20)

### Only small droplets contribute to "accuracy" portion

[Compare](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreProcessorFruits.cs#L30) [stable.](https://github.com/peppy/osu-stable-reference/blob/master/osu!/GameplayElements/Scoring/Processors/ScoreProcessorFruits.cs#L38)

### Bananas now give 200 bonus

[Compare stable.](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreChangeFruits.cs#L21-L22)

A notable edge case difference here is that on beatmaps that do not have any fruits, stable score V2 would just [refuse to give any points whatsoever](https://github.com/peppy/osu-stable-reference/blob/dc0994645801010d4b628fff5ff79cd3c286ca83/osu!/GameplayElements/Scoring/Processors/ScoreProcessorFruits.cs#L27-L28), even if bananas were caught. This matters for beatmaps that only consist of banana showers, such as [this one](https://osu.ppy.sh/beatmapsets/23210#fruits/80895). lazer will instead grant 1M score for free and then tack on points from bananas caught onto that.

---

All of the above was experimentally and organoleptically examined, cross-checked, and compared against stable. Comparisons were generally not easy, because replay playback consistency is not guaranteed, but I believe the implementation to now be matching 1:1.

## Score conversion

To match the modified scoring algorithm, the conversion code from score V1 was also adjusted to match.

I started off from a very simple version of it, but on the resulting spreadsheet I found it to have very wrong results for some beatmaps. Further investigation showed that due to differences in rate of growth between score V1 and score V2, unique to catch, the previous method of estimation of combo portion was not suitable in specific circumstances.

To spare the reader's eyes, the derivation of the algorithm implemented in this PR is explained [in this gist](https://gist.github.com/bdach/414d5289f65b0399fa8f9732245a4f7c#file-conversion-algorithm-md). Enter at your own risk.

The resulting spreadsheet showing the impact of the modified score conversion can be found [here](https://docs.google.com/spreadsheets/d/1nFFByY5Yyfp1eZkQq_-uCoKdkPGnFK9-zzxd2iW2MBo/edit#gid=1340545131). The analysis of selected outliers on that spreadsheet can be found [in this gist](https://gist.github.com/bdach/414d5289f65b0399fa8f9732245a4f7c#file-spreadsheet-analysis-md).

All replays used in testing above are here: [replays.zip](https://github.com/ppy/osu/files/13846683/replays.zip)

Notably, some score conversions were and continue to be visibly broken, and are not fixed due to overarching concerns that affect all rulesets, not only catch. I may attempt to fix them separately, but that might require more time.

### Legacy scoring attributes need to be adjusted for catch ruleset

The specific change that forces this is the change of bonus given by bananas. Because `BonusScoreRatio` is a ratio of bonus given by standardised scoring to ratio of bonus given by stable score V1 bonus, changing the number of points given by each banana obviously changes the ratio itself.

The simplest and safest way to resolve this would be to do a full run `osu-difficulty-calculator` over catch beatmaps and converts. However, a riskier but quicker option would be to do this manually, leveraging the fact that the change to `BonusScoreRatio` is simple - because bonus score per banana rose (200 / 50) = fourfold, it may be enough to just execute something along the lines of **(WARNING, UNTESTED, CHECK FIVE TIMES BEFORE EXECUTING)**:

```sql
UPDATE `osu_beatmap_scoring_attribs`
SET `legacy_bonus_score_ratio` = 4 * `legacy_bonus_score_ratio`
WHERE `mode` = 2;
```

### Stable scores need to be reimported

Because this pull changes how both accuracy and total score are computed when importing stable scores, a full reimport of stable scores for catch is required.

---

cc @Firuqwq @bastoo0 @SecreOsu @Kimitakari @Nirfu as potentially interested in cross-checking / testing this

### CODE REVIEW

File: osu.Game/Database/StandardisedScoreMigrationTools.cs [hunk 1]
```diff
@@ -461,6 +482,94 @@ private static long convertFromLegacyTotalScore(ScoreInfo score, LegacyBeatmapCo
...
+            if (assumedLengthOfRemainingCombos > 0)
+            {
+                while (remainingCombo > 0)
+                {
+                    int comboLength = Math.Min(assumedLengthOfRemainingCombos, remainingCombo);
+
+                    remainingCombo -= comboLength;
+                    totalDroppedScore += estimateDroppedComboScoreAfterMiss(comboLength);
+                }
+            }
```
Comment (bastoo0) [inline osu.Game/Database/StandardisedScoreMigrationTools.cs:529]: As discussed on Discord, you can remove the while loop here, if that even matters.
```suggestion
           if (assumedLengthOfRemainingCombos > 0)
        {
            // How many combos we have to evaluate (still estimate)
            int combosCount = (int)Math.Floor((double)remainingCombo / assumedLengthOfRemainingCombos);
            
            // We may still have objects remaining after the last combo evaluation, if remaining objects can't be evenly split by the assumed length of remaining combos
            int remainingObjects = remainingCombo - (combosCount * assumedLengthOfRemainingCombos);

            totalDroppedScore += combosCount * estimateDroppedComboScoreAfterMiss(assumedLengthOfRemainingCombos);
            
            if (remainingObjects > 0)
            {
                totalDroppedScore += estimateDroppedComboScoreAfterMiss(remainingObjects);
            }
        }
```

Comment (bdach) [reply osu.Game/Database/StandardisedScoreMigrationTools.cs:529]: Yeah I'll gauge the general reaction to *vaguely points at everything else in this diff* and if it doesn't get kicked to the curb I'll apply this change and re-validate that results don't change.

Comment (bdach) [reply osu.Game/Database/StandardisedScoreMigrationTools.cs]: Have applied.

!diffcalc
GENERATORS=score
RULESET=catch

**Comment by bdach** [comment]
!diffcalc
GENERATORS=score
RULESET=catch

**Review by peppy — APPROVED** [review]

**Pull request merged**